### PR TITLE
Foundation: Support contextual theming and styles

### LINKED ISSUE #5721: Foundation: Component Creation
## Component Details

There is a need to create a foundation package that improves and extends the existing `styled` functionality that OUFR provides. This new Foundation package has the following goals:

| Goal      | Rationale     | Goal Currently Met | Status |
|-----------|----------|----------|--------|
| Centralize style processing and merging. | Reduce error-prone component boilerplate of using, merging and applying theming and styling. | :white_check_mark: | |
| Create separation of concerns for improved maintenance and scalability: state, view and styling. | Improve ease of testability, maintenance, reusability and scalability for component implementations. | :white_check_mark: | | 
| Introduce optional state component. | State handles statefulness and passing transformed props and state to view. State components do not generate JSX and simply call a view function in their render function. State does not need to be concerned with using or managing styling props. Unit tests can be concerned with testing statefulness without worrying about rendered output. | :white_check_mark: |
| Automatic processing and merging of styles. | Foundation handles passing component props and state component output through styling and passing resulting styled output to view. | :white_check_mark: |
| Introduce core view component. | Views are pure functions whose input are all props. Since views do not have state their unit test coverage simply becomes snapshot tests covering prop permutations. Components without states are effectively stateless components where styling is automatically applied to component props. | :white_check_mark: | |
| Centralize common controlled state patterns, like toggle. | Reduce the need for components to duplicate common state logic patterns. Improve reliability of controlled vs. uncontrolled behavior. | :white_check_mark: | |
| Maintain backwards compatibility. | For existing components and for consumers who wish to continue creating combined state/view components, continue to offer existing `styled` implementation in Foundation package. | :white_check_mark: | Resolution: Continue to support `styled` separately until deprecated. |
| Be agnostic of component library and styling approach and impose as minimal structure on styling and component surface as possible. Do not depend on any other OUFR package. | Allow adoption by component libraries other than OUFR. | :white_check_mark: | |
| Eliminate state component layer if at all possible, leaving simply composed component and theme consuming layers. | Maximize scalability and perf. | :white_check_mark: | State component layer ultimately not eliminated. See #5799 for rationale. |

#### createComponent

createComponent is the core function for Foundation, consuming the view, state and styles objects to create an HOC that handles all styling and props processing.

image

__Arguments__

| Name      | Type     | Description |
|-----------|----------|-------------|
| options | `IComponentOptions` | See `IComponentOptions` below. |
| providers | `IStyleProviders` | Functions needed by Foundation to process and merge style sets. |
| StateComponent | `IStateComponent` | Optional state component with additional props provided by Foundation, such as `renderView` prop. |

__IComponentOptions__

| Name      | Type     | Description |
|-----------|----------|-------------|
| displayName | `string` | Display name to identify component in React hierarchy. |
| styles | `IStylesProp<TViewProps, TStyleSet>` | Styles prop to pass into component. |
| view | `(props: IViewComponentProps<TViewProps, TProcessedStyledSet>) => JSX.Element` | Functional React view component. |
| statics | `TStatics` | Optional static object to pass into constructed component. |

__Types__

Types are used to eliminate dependencies on specific styling library, allowing Foundation to have no other OUFR dependencies. This requires the consuming component library to provide these types.

| Type     | Description |
|-----------|----------|
| `TComponentProps` | A styleable props interface for the created component. This is the API surface for the created component. |
| `TViewProps` | The props specific to the view, including processed properties outputted by optional state component. If state component is not provided, `TComponentProps` is the same as `TViewProps`. |
| `TStyleSet` | The type for styles properties. |
| `TProcessedStyleSet` | The type provided by `mergeStyleSets` provider after processing `TStyleSet` and provided to views. |
| `TTheme` | The type for contextual and prop themes. |

__Component Library Interface__

It is recommended that the component library provide an interface file that passes in global types (such as OUFR's `ITheme` and `IProccessedStyleSet`) to Foundation so that the interfaces for components are simplified. For example:

```
export type IViewProps<TProps, TStyleSet extends IStyleSet<TStyleSet>> = IViewProps<TProps, IProcessedStyleSet<TStyleSet>>;
export type IStyleableComponent<TProps, TStyleSet> = IStyleableComponent<TProps, TStyleSet, ITheme>;
```

### Imports

As one of the primary goals of Foundation is to be agnostic of styling and component libraries, Foundation should have no other OUFR dependencies.

### Intended Package
`@uifabric/foundation`

### Code mockup/example

The API approach is similar to `styled`, the main differences being the explicit callout of `state` and `view` arguments.

__Code Sample__

```
import { createComponent } from '../../Foundation';
import { CollapsibleSectionView } from './CollapsibleSection.view';
import { CollapsibleSectionState } from './CollapsibleSection.state';
import { getStyles as styles } from './CollapsibleSection.styles';
import {
  ICollapsibleSectionProps,
  ICollapsibleSectionControlledProps,
  ICollapsibleSectionViewProps,
  ICollapsibleSectionStyles
} from './CollapsibleSection.types';

export const CollapsibleSection: React.StatelessComponent<ICollapsibleSectionProps> = createComponent<
  ICollapsibleSectionProps,
  ICollapsibleSectionViewProps,
  ICollapsibleSectionStyles
>(
  {
    displayName: 'CollapsibleSection',
    view: CollapsibleSectionView,
    styles
  },
  CollapsibleSectionState
);
```

### Component Ownership
This will be a core Fabric package with initial implementation and support provided by the core Fabric team (main POC @JasonGore.)

## TODOs
| TODO      | Complete     | PR |
|-----------|----------|-------------|
| Is it possible to combine state objects within HOC component to eliminate a separate StateComponent level in the React hierarchy?  |  :white_check_mark: | #5799  |
| Chainable state / autocontrolled patterns. Can we have predefined autocontrolled patterns like 'toggle' so that such state components don't have to be duplicated for each component?  | :white_check_mark: | #5799 |
| A pattern for making controlled vs. uncontrolled components. Avoid duplicating sources of truth when consuming component provides controlled props. Minimize logic required to get source of truth with core components ideally not storing state at all. Potential ideas include a controlled state object that is the source of truth for when consumer does not define a value for controlled props.  TextField as potential conversion example (default props vs. controlled props) |  :white_check_mark: | #5799  |
| Add support for contextual theming and remove `getTheme` provider.  | :white_check_mark: | #5867 |
| Resolve (requiring new TS ver) or find potential workarounds for TypeScript-related issues: Having to export unused vars (can't export unimported external types.) Needing two implementations of createComponent for type safety (need conditional types.) Superfluous types (love to be able to have components extend types so that users don't need to define as many types.)  | :white_check_mark: | #6474  |
| Update `create-component` script and [New Component wiki](https://github.com/OfficeDev/office-ui-fabric-react/wiki/New-Components). | :white_check_mark: | #7569 |
| Support React 16 context. (See #5701)  | :white_check_mark: | #6268 |
| Consolidate all copies of createComponent into Foundation.  |  :white_check_mark: | #5693 |
| Add type safety and formalize state/view relationship and typing as well as props merging and handling  | :white_check_mark: | #5693 |

## Steps

- [x] is themable (base/styles)
- [ ] has unit tests
- [ ] has snapshot tests
- [ ] has visual tests
- [x] has documentation and examples
- [ ] Design Review
- [ ] API review
- [ ] Public Preview
- [ ] Ready for Publishing

### PR DESCRIPTION
**Pull request by JasonGore**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #5721
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Add support for contextual theming, styles, settings and scoped settings to createComponent. Create examples to demo.

There's a bit of engineering debt accruing here but seeing as we wanted something for Friday and I'm on Shield, we can either leave this open until issues are addressed or merge and follow up.

context

#### Focus areas to test

Tests to be written (aforementioned engineering debt.)

### CODE REVIEW

File: packages/experiments/src/Foundation.ts [hunk 1]
```diff
@@ -85,7 +94,17 @@ export function createComponent<
...
   >(options, providers, state);
 }
+
+// TODO: remove any if possible
+// tslint:disable-next-line:no-any
+function getCustomizations(displayName: string, context: IContextCustomization): any {
+  // TODO: do we want field props? should fields be part of IComponentOptions and used here?
+  // TODO: should we centrally define DefaultFields? (not exported from styling)
+  const DefaultFields = ['theme', 'styles'];
+  return Customizations.getSettings(DefaultFields, displayName, context.customizations);
```
Comment (dzearing) [inline packages/experiments/src/Foundation.ts]: defaultfields are default... some components, like Layer, can mark additional fields as customizable (such as "hostId"). How is this piped in?

File: packages/experiments/src/components/CollapsibleSection/CollapsibleSection.ts [hunk 2]
```diff
@@ -21,3 +21,13 @@ export const CollapsibleSection: React.StatelessComponent<ICollapsibleSectionPro
...
 );
+
+// TODO: This is only here for testing createComponent and should be removed before promoting to production
+export const CollapsibleSectionStateless: React.StatelessComponent<ICollapsibleSectionProps> = createStatelessComponent<
+  ICollapsibleSectionViewProps,
+  ICollapsibleSectionStyles
+>({
+  displayName: 'CollapsibleSection',
+  view: CollapsibleSectionView,
+  styles
```
Comment (dzearing) [inline packages/experiments/src/components/CollapsibleSection/CollapsibleSection.ts]: customizableFields: ['foo', 'bar', 'baz']; maybe

File: packages/experiments/src/Foundation.ts (hunk 1 above)
Comment (JasonGore) [reply packages/experiments/src/Foundation.ts]: Yeah, I wanted to talk to you about that. I have the TODO on line 106 for exactly that. 😄

File: packages/experiments/src/components/CollapsibleSection/CollapsibleSection.ts (hunk 2 above)
Comment (JasonGore) [reply packages/experiments/src/components/CollapsibleSection/CollapsibleSection.ts]: Yep, that's what I was thinking, assuming we still needed to pipe them in (tied to other comment)

**Pull request merged**